all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+
+/// Users might want to be able to pause the handing-out of slots as an effective way of pausing their workers.
+/// We can provide an implementation for this that wraps their implementation, or one of the defaults we provide.
+struct PauseableSlotSupplier<T: SlotKind> {
+  inner: Arc<dyn SlotSupplier<SlotKind=T>>,
+  paused: AtomicBool,
+  // ... details
+}
+impl<T: SlotKind> PauseableSlotSupplier<T> {
+  pub fn new(supplier: impl SlotSupplier<SlotKind=T>) -> Self { /* ... */ }
+  pub fn pause(&self) { /* ... */ }
+  pub fn resume(&self) { /* ... */ }
+}
+impl<T: SlotKind> SlotSupplier for PauseableSlotSupplier<T> {
+  // ... implementation which checks the paused flag before handing out a slot
+}
+```
+
+## Workflow cache management
+
+The workflow cache is maintained independently of the number of slots in use. There could be 0 slots in use and 100
+workflows cached, and that's a totally normal scenario. We might also choose to make this user-configurable, though we
+certainly don't need to have that right out of the gate.
+
+Keeping our fixed-size implementation to start is fine, because we won't ask the user's slot provider for a slot unless
+there is room in the cache (either additional capacity, or an idle cached workflow that can be evicted). This avoids
+a potential problem where we ask for a slot, they provide one, but there's nothing we can evict at the moment.
+
+If and when we do introduce user-configurable caching, it's likely that they would want the same thing to be responsible
+for both slots and cache management, since the two are closely related.
+
+A possible future:
+
+```rust
+trait WorkflowCacheSizer {
+  /// Return true if it is acceptable to cache a new workflow. Information about already-in-use slots, and just-received
+  /// task is provided. Will not be called for an already-cached workflow who is receiving a new task.
+  /// 
+  /// Because the number of available slots must be <= the number of workflows cached, if this returns false
+  /// when there are no idle workflows in the cache (IE: All other outstanding slots are in use), we will buffer the
+  /// task and wait for another to complete so we can evict it and make room for the new one.
+  fn can_allow_workflow(&self, slots_info: &WorkflowSlotsInfo, new_task: &WorkflowSlotInfo) -> bool;
+  /// Called when deciding what workflow should be evicted from the cache. May return the run id of a workflow to evict.
+  /// If None, or an invalid run id is returned, the default LRU eviction / strategy is used. The id returned must
+  /// be one of the workflows provided in `evictable` (the ones for which there is no actively processing WFT).
+  fn eviction_hint(&self, evictable: &[WorkflowSlotInfo]) -> Option<&str>;
+  /// Called when a workflow is evicted from the cache
+  fn evicted_workflow(&self, evicted: &WorkflowSlotInfo);
+  /// If there is a fixed maximum number of cached workflows, return it.
+  fn max(&self) -> Option<usize>;
```

Comment:
What is the value of this call on the interface? Isn't a fixed-max an implementation of this interface not a part of every implementation?

Reply:
That's why it's optional - implementations that know it return it, if not they return None. We can emit a metric about total cache size if it's returned here.

Reply:
But I think the impl can expose its own metric about total cache size if it cares to, the system doesn't need to care whether a max exists or not.

Reply:
I think it's nice to have that metric be standardized when it is available. It's another thing support can look at when available, etc.

Reply:
I think utilities that help code reuse may make sense, but having every interface implement something only a few need just to show a metric that some may not have seems unnecessary. Just makes for larger implementation surface.

Reply:
I think metric emitting should be owned by the slot supplier itself not based on what interfaces it implements

Reply:
> I think metric emitting should be owned by the slot supplier itself not based on what interfaces it implements

The supplier is free to emit whatever it likes - there just happen to be a couple known ones that we already emit, and I don't see why we wouldn't want to always emit those under a consistent name if the user happens to be able to provide it.

Reply:
> that we already emit

We don't already emit max slots in any SDK do we?

> see why we wouldn't want to always emit those under a consistent name if the user happens to be able to provide it.

 Consider I share a slot provider that implements this interface across 3 workers because I want to restrict the total number of slots used across these 3 workers bounded. Now since your atomically emitting the metric instead of letting the slot supplier do it you are emitting 3x the value. I think we should keep slot supplier metrics in the slot supplier and add utilities to help code reuse if we see any.

Reply:
> We don't already emit max slots in any SDK do we?
Good point lol, we don't.

The example with sharing implementation confusion is legit. Fair enough, we can put these metrics behind functions a user is welcome to call if they provide the same functionality.

Reply:
I would like to not emit a metric that doesn't have a valid value. What would you emit for available slot metric for implementations that have no concept of available slots? No integer is acceptable here. Slot suppliers should own their metrics, and there is only one slot supplier we will provide where available slots makes sense (the default one).